VSCopy: Enable to copy from all shards in either a specified keyspace or all keyspaces

yoheimuta opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->

This PR removes the two checks by which VTGate refuses a copy request lacking either shard or keyspace.

Let's say you have two keyspaces named `ks` and `ks2,` and each keyspace includes two shards called `-80` and `80-.`
Here's the table of the input and expected output:

| gtid  | shard | keyspace | copy targets |
|-------|-------|----------|--------------|
| empty | empty | /.*    | `ks.-80`, `ks.80-`, `ks2.-80`, `ks2.80-`       |
| empty | empty | ks       | `ks.-80`, `ks.80-`       |

## Related Issue(s)

<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->

Fixes https://github.com/vitessio/vitess/issues/11886

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

yoheimuta commented on go/vt/vttest/local_cluster.go:
```diff
@@ -151,20 +151,20 @@ type Config struct {
+		if err != nil {
+			return err
+		}
+		schemaDir = tempSchemaDir
```
The main change in this file is the above lines. 
The below is a trivial diff.

mattlord replied:
OK. WHY are we doing this? 🙂

yoheimuta commented on go/vt/vtgate/endtoend/row_count_test.go:
```diff
@@ -24,13 +24,15 @@ import (
 	conn, err := mysql.Connect(ctx, &vtParams)
 	require.NoError(t, err)
 	defer conn.Close()
+	utils.Exec(t, conn, "use ks")
```
This use statement is now necessary because the number of keyspace is multiple. 3afc402

mattlord commented:
Hi @yoheimuta,

Thank you for another PR! I just wanted to note that before we merge this, we need to catch up on the vtgate VStream RPC docs: https://github.com/vitessio/website/pull/1216

We can add info there about the new ability to copy not only individual tables, but all tables in a keyspace or all tables in all keyspaces. I'll link this PR to that docs PR for now. You should be able to push directly to that PR but if not let me know and I can get you setup with the website repo. 

We're not supposed to merge PRs w/o any accompanying docs but we did, and I don't want to fall further behind on that front.

(There's another tangentially related PR here too: https://github.com/vitessio/vitess/pull/11771)

Thanks!

yoheimuta commented:
Thank you for telling me! I'm checking out the docs PR.

yoheimuta commented:
> You should be able to push directly to that PR but if not let me know and I can get you setup with the website repo.

@mattlord I couldn't push directly to the PR.
Can you please look into the access rights?

```
(base) ❯ git push --set-upstream origin vstream_docs
ERROR: Permission to vitessio/website.git denied to yoheimuta.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

mattlord commented on go/vt/vtgate/vstream_manager.go:
```diff
@@ -179,13 +179,8 @@ func (vsm *vstreamManager) resolveParams(ctx context.Context, tabletType topodat
+	// that has an empty keyspace.
 	if len(vgtid.ShardGtids) == 1 && vgtid.ShardGtids[0].Keyspace == "" {
-		if vgtid.ShardGtids[0].Gtid != "current" {
-			return nil, nil, nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "for an empty keyspace, the Gtid value must be 'current': %v", vgtid)
-		}
 		keyspaces, err := vsm.toposerv.GetSrvKeyspaceNames(ctx, vsm.cell, false)
 		if err != nil {
 			return nil, nil, nil, err
```
I would prefer the request be more explicit as it will likely generate A LOT of load and could have a big impact on production. I think it would also be nice if the keyspace wildcard selection lined up with the table wildcard selection in `binlogdata.Filter.Rule.Match`, which is `/.*` as a valid regular expression is accepted if the string has a `/` prefix. So I think accepting valid regular expressions for the keyspace would be much nicer: `/.*`.

yoheimuta replied:
@mattlord It makes sense in terms of the difference in impact between "streaming" and "copy."
Accepting the keyspace wildcard sounds useful. With this feature, we don't have to accept the request with an empty keyspace and empty gtid.

Do you think the following specifications are still required?

- A special request flag, for example, "EnableEmptyKeyspace" or "EnableWildcardKeyspace" in VStreamFlags.
   - I think only the keyspace wildcard like `/.*` is enough.
- The valid combination of an empty keyspace and "current" gtid.
   - While the combination of an empty keyspace and the gtid other than "current" is invalid. It complicates the code, but not so much.
   - Some may think it's better to stop accepting this input for the sake of simplicity at this moment.

As such, I plan to update accepting the keyspace wildcard and preserve the previous behavior as much as possible like the below table. What do you think?

| gtid  | keyspace | valid or invalid(=noop) |
|-------|-------|--------------|
| "" | ""    | invalid because of no literal match. No change.     |
| "current" | ""       | valid because of the compatibility. No change.      |
| "" | "/.*"       | valid if found because of wildcard match. New behavior.      |
| "current" | "/.*"       | valid if found because of wildcard match. New behavior.      |
| "" | "/ks.*/"       | valid if found because of wildcard match. New behavior.      |
| "current" | "/ks.*/"       | valid if found because of wildcard match. New behavior.      |

mattlord replied:
I think the explicit keyspace wildcard is enough and we don't need a new flag around that.

Now that I understand the usage of `current` better, I think we have to keep it in place for compatibility sake. What do you think @rohit-nayak-ps ?

yoheimuta replied:
Hi @rohit-nayak-ps, I'd appreciate your feedback. Let me know if anything is missing.

rohit-nayak-ps replied:
@yoheimuta, sorry for the delay. I am fine with the approach outlined above.

We should add a (or modify the existing) e2e test to confirm that resharding will continue to work if multiple keyspaces are being streamed.  We have code (see the logic around `journaler` in `vstream_manager`), which waits for participating shards to converge, before automatically continuing to stream from the new shards.  I think it should be fine, when there are multiple keyspaces, but confirming with a test will be useful.

yoheimuta replied:
Thank you for your confirmation!
I'm looking into e2e test too.

yoheimuta replied:
Hi 👋 

@mattlord I implemented the keyspace wildcard selection by [d83daaf](https://github.com/vitessio/vitess/pull/11909/commits/d83daaf0a244aac6c69d34d709c5fe76e9cdc640) and [c39af7c](https://github.com/vitessio/vitess/pull/11909/commits/c39af7ce7379d93b5330ad78919cfc0b908e7d5a).

@rohit-nayak-ps I added the e2e test by [71e1681](https://github.com/vitessio/vitess/pull/11909/commits/71e16813dc87d980e80a51636874124e5338a8f2) and [436568d](https://github.com/vitessio/vitess/pull/11909/commits/436568de9f2fe7d5aedabdf86dea6108ce3e2243)

And I think the CI failure is caused by some unreliable tests because the failing test changes every time we run it and it passes when we run it locally.

rohit-nayak-ps replied:
Thanks @yoheimuta, we will review it soon. Yes, that CI failure is also happening on other PRs and I have asked the related team to look at it.

mattlord commented on go/vt/vtgate/vstream_manager_test.go:
```diff
@@ -886,16 +886,37 @@ func TestResolveVStreamParams(t *testing.T) {
 	}, {
 		input: &binlogdatapb.VGtid{
-			ShardGtids: []*binlogdatapb.ShardGtid{{}},
+			ShardGtids: []*binlogdatapb.ShardGtid{{
+				Keyspace: "TestVStream",
+			}},
```
These test additions are unrelated to this PR, or no? I would think we would test the new behavior, which would mean NOT specifying the shards and now not getting an error?

yoheimuta replied:
@mattlord I thought they were related to this PR.
Technically, I didn't add new test cases (input & output). I changed the expected output in the following existing two cases:

1. When the input is `ShardGtids: []*binlogdatapb.ShardGtid{{}},`
     - prev output: Returning an error
     - new output: Returning all shards in all keyspaces without any errors. 
     - Moving the case code to the special-case tables because output is too big. https://github.com/vitessio/vitess/pull/11909/files#diff-3e2894943e5628aa7d63298e06c321fe82602d18693b2d5e99541fafd2c01395R1021
2. When the input is `ShardGtids: []*binlogdatapb.ShardGtid{{Keyspace: "TestVStream"}},`
     - prev output: Returning an error
     - new output: Returning all shards in the TestVStream keyspace without any errors. 
     - https://github.com/vitessio/vitess/pull/11909/files#diff-3e2894943e5628aa7d63298e06c321fe82602d18693b2d5e99541fafd2c01395R893

mattlord replied:
Here as well I think that comments noting what we're actually testing, how, and why are helpful. The future reader won't easily see comments here on the PR. And I don't know that the existing code is very poorly commented, but I want to do better moving forward. 🙂

yoheimuta replied:
Added by [3b41d76](https://github.com/vitessio/vitess/pull/11909/commits/3b41d767bd0e6a9720bcaaa5ed9e3b7591540da3).

mattlord commented on go/vt/vtgate/vstream_manager_test.go:
```diff
@@ -987,17 +1008,34 @@ func TestResolveVStreamParams(t *testing.T) {
-		ShardGtids: []*binlogdatapb.ShardGtid{{
-			Gtid: "current",
-		}},
+	specialCases := []struct {
+		input  string
+		output string
+	}{
+		{
+			input:  "current",
+			output: "current",
+		},
```
Do we still want/need the special `current` keyword and associated behavior? That seemed to be tied to "no keyspace value", which you altered here. Perhaps it makes sense to keep `current` thought if we use `/.*` instead of no value for the keyspace wildcard.

yoheimuta replied:
I thought it was worth remaining the previous typical request to explicitly verify one of the main use cases even though my change tried to leave the case intact. But removing the irrelevant code in the latest code also makes sense. If I double down, I should add a comment about the background.
Anyway, as you mentioned, I'm revisiting this topic after settling the keyspace wildcard.

mattlord replied:
Yeah, I think you're right. Let's get @rohit-nayak-ps's opinion too though.

mattlord commented on go/vt/vtgate/vstream_manager.go:
```diff
@@ -179,31 +180,51 @@ func (vsm *vstreamManager) resolveParams(ctx context.Context, tabletType topodat
-			if sgtid.Gtid != "current" {
-				return nil, nil, nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "if shards are unspecified, the Gtid value must be 'current': %v", vgtid)
+			if sgtid.Gtid != "current" && sgtid.Gtid != "" {
+				return nil, nil, nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "if shards are unspecified, the Gtid value must be 'current' or empty: %v", vgtid)
```
Suggestion:
```
return nil, nil, nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "if shards are unspecified, the Gtid value must be 'current' or empty; got: %v", vgtid)
```

yoheimuta replied:
Applied by [d730563](https://github.com/vitessio/vitess/pull/11909/commits/d730563f5a5c5d7dcd72f6ca27060898549aa4ed).

mattlord commented on go/vt/vtgate/vstream_manager_test.go:
```diff
@@ -987,17 +1021,47 @@ func TestResolveVStreamParams(t *testing.T) {
-		ShardGtids: []*binlogdatapb.ShardGtid{{
-			Gtid: "current",
-		}},
+	specialCases := []struct {
```
I'm running out of steam on the review as I have to try to figure out what each test is doing, how, and why. :-) What is this test case about? I don't think the comment above is accurate, is it? In the test we then seem to be checking for a length of 8. Why? I know that the reader can dig into the code to answer these questions, but it's a lot to place on the reader.

yoheimuta replied:
I overlooked updating the comment above 🙇‍♂️ , so I made the necessary corrections in [d7169ec](https://github.com/vitessio/vitess/pull/11909/commits/d7169ec59f3e28e6d0320525a4cf2be3f8878a0c). 
This commit also clarifies the intent of the original code.

mattlord reviewed (commented):
In general this looks great! For the most part my requests revolve around commenting the tests. It took me a LONG time to review the relatively small PR as I had to try and figure out what was being done (mostly told by the code itself), why it was being done (what are we trying to test/confirm), and how we were then confirming it. It's a lot for each reader to do each time they look at the tests. I know that the existing code/tests were very poorly commented, but I'm trying to be better going forward. I'm happy to do another review quickly.

In short, it's very challenging and time consuming to review a test for correctness and validity when the writer's intentions are not clearly and explicitly stated. 

Thank you! ❤️

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+func testVStreamCopyMultiKeyspaceReshard(t *testing.T, baseTabletID int) *numEvents {
+	defaultCellName := "zone1"
+	allCells := []string{"zone1"}
+	allCellNames = "zone1"
```
Fixed by [ca2dafd](https://github.com/vitessio/vitess/pull/11909/commits/ca2dafdee25521dbb0bf261f793cdb1699391ed4).

Note: My choice of this particular coding was in the expectation that the same existing code would serve as a sensible role model.
- https://github.com/vitessio/vitess/blob/adb2535cf79926d2d9ecf9710a280d657103f74a/go/test/endtoend/vreplication/vstream_test.go#L233-L235

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+
+	require.NotNil(t, vc)
+	defaultReplicas = 0 // because of CI resource constraints we can only run this test with primary tablets
+	defer func() { defaultReplicas = 1 }()
```
Thank you for your suggestion! Fixed by [24e69b9](https://github.com/vitessio/vitess/pull/11909/commits/24e69b9499b6ae7329bf777b5c4a9a77b46f7749).

Note: My choice was in the expectation that the same existing code would serve as a practical role model.
- https://github.com/vitessio/vitess/blob/adb2535cf79926d2d9ecf9710a280d657103f74a/go/test/endtoend/vreplication/vstream_test.go#L238-L240

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
 	return &ne
 }
 
+// tests the VCopy with multiple keyspaces and resharding
```
Added by [7e67f27](https://github.com/vitessio/vitess/pull/11909/commits/7e67f27e45e9fb9484f12506e653755c52675bf7).
I didn't realize that the comments in the test code were important for this project as far as I referred to this file. Thank you for clarifying your expectation that we should write informative comments in our code going forward. This will help future contributors and me.

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+
+	filter := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match: "/customer.*/",
```
Added by [dc98a42](https://github.com/vitessio/vitess/pull/11909/commits/dc98a420e458316c37398389d759b69e02c04967)

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+				return
+			}
+			id++
+			time.Sleep(1 * time.Second)
```
I referred to a similar test case that also verifies the VStream API during resharding and confirmed that 1 second works for this case as well. 

https://github.com/vitessio/vitess/blob/adb2535cf79926d2d9ecf9710a280d657103f74a/go/test/endtoend/vreplication/vstream_test.go#L292

If you feel that we need more time for this, let's discuss the rationale in the chat you set up for us. Thanks!

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+		}
+	}()
+
+	ticker := time.NewTicker(1 * time.Second)
```
NOTE: I think we need to address https://github.com/vitessio/vitess/pull/11909#discussion_r1103334686 first.

mattlord replied:
OK. Not a big deal, was mostly curious.

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+			break
+		}
+	}
+	return &ne
```
Fixed by [6a791ba](https://github.com/vitessio/vitess/pull/11909/commits/6a791ba7c9806237446743f65dffdb0af54a9b3f) because I was just adapting to the existing code (introduced https://github.com/vitessio/vitess/commit/88ad81f30a2dc0f68eaba3b2195a56eef9e0557c#diff-bb66213b66b91e5690c6ab2df09eb1ea7f48214225497046f8ea012f1bf9bdb4R368) and thought this practice was unusual but harmless in most cases.

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -375,6 +380,132 @@ func testVStreamStopOnReshardFlag(t *testing.T, stopOnReshard bool, baseTabletID
+			Match: "/customer.*/",
+		}},
+	}
+	flags := &vtgatepb.VStreamFlags{HeartbeatInterval: 3600}
```
Deleted by [c2138b5](https://github.com/vitessio/vitess/pull/11909/commits/c2138b59a4ceb71e5a3e162c8c670c2892c3b111) because the similar tests set this flag without any explanation and both tests passed locally regardless of this flag, so I set the same flag just in case. Thank you for telling me!

- https://github.com/vitessio/vitess/blob/adb2535cf79926d2d9ecf9710a280d657103f74a/go/test/endtoend/vreplication/vstream_test.go#L85
- https://github.com/vitessio/vitess/blob/adb2535cf79926d2d9ecf9710a280d657103f74a/go/test/endtoend/vreplication/vstream_test.go#L279

yoheimuta commented on go/test/endtoend/vreplication/vstream_test.go:
```diff
@@ -406,3 +537,16 @@ func TestVStreamWithKeyspacesToWatch(t *testing.T) {
+	ne := testVStreamCopyMultiKeyspaceReshard(t, 3000)
+	log.Infof("ne=%v", ne)
+	require.Equal(t, int64(0), ne.numJournalEvents)
+	require.NotZero(t, ne.numRowEvents)
+	require.NotZero(t, ne.numShard0BeforeReshardEvents)
+	require.NotZero(t, ne.numShard0AfterReshardEvents)
+	require.NotZero(t, ne.numLessThan80Events)
+	require.NotZero(t, ne.numGreaterThan80Events)
+	require.NotZero(t, ne.numLessThan40Events)
+	require.NotZero(t, ne.numGreaterThan40Events)
```
I added an additional test to ensure that we do not lose any events or receive duplicates when resharing the keyspace, as seen in commit 16f10b3.

I don't think that this implementation is particularly flaky for the following reasons:

- If the CI is slow, vc.AddKeyspace will fail due to a timeout in the first place.
- It seems reasonable to expect four insert statements that increment numLessThan40Events, numGreaterThan40Events, and numShard0AfterReshardEvents to complete within 60 seconds.

yoheimuta commented on go/vt/vtgate/endtoend/misc_test.go:
```diff
@@ -55,6 +56,16 @@ func TestCreateAndDropDatabase(t *testing.T) {
 
+	// cleanup the keyspace from the topology.
+	defer func() {
+		// the corresponding database needs to be created in advance.
+		// a subsequent DeleteKeyspace command returns the error of 'node doesn't exist' without it.
```
Fixed by [6910cbc](https://github.com/vitessio/vitess/pull/11909/commits/6910cbce207eb3e76f88400fd979be17db90720e)

yoheimuta commented on go/vt/vtgate/endtoend/misc_test.go:
```diff
@@ -55,6 +56,16 @@ func TestCreateAndDropDatabase(t *testing.T) {
+	defer func() {
+		// the corresponding database needs to be created in advance.
+		// a subsequent DeleteKeyspace command returns the error of 'node doesn't exist' without it.
+		_ = exec(t, conn, "create database testitest")
```
I had no reason to use this condition because the `testitest` database was only used in `TestCreateAndDropDatabase`, as shown below:

https://github.com/vitessio/vitess/blob/a4f1d2f744502a13f7f7b0fbef136f9398fc72b8/go/vt/vtgate/endtoend/misc_test.go#L61-L66

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
 	}
 }
 
+func TestVStreamCopyWithoutKeyspaceShard(t *testing.T) {
```
Added by [c70fa39](https://github.com/vitessio/vitess/pull/11909/commits/c70fa39180d8a1162fed1a036585fe7b1a750262)

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+	conn, err := mysql.Connect(ctx, &vtParams)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Replaced by [e8f0e60](https://github.com/vitessio/vitess/pull/11909/commits/e8f0e60d4b232bca4c257fbf661bb442ab84c390)

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+	}
+	flags := &vtgatepb.VStreamFlags{}
+
+	expectedKs1EventNum := 2 /* num shards */ * (9 /* begin/field/vgtid:pos/4 rowevents avg/vgitd: lastpk/commit) */ + 3 /* begin/vgtid/commit for completed table */ + 1 /* copy operation completed */)
+	expectedKs2EventNum := 2 /* num shards */ * (6 /* begin/field/vgtid:pos/1 rowevents avg/vgitd: lastpk/commit) */ + 3 /* begin/vgtid/commit for completed table */ + 1 /* copy operation completed */)
+	expectedFullyCopyCompletedNum := 1
```
Thanks! Applied by [cdc4077](https://github.com/vitessio/vitess/pull/11909/commits/cdc4077fcb7c28071fe2e570ddec7345c7b381b4)

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+			},
+		},
+		{
+			name: "copy from all shards in the keyspace",
```
Thanks! [6b2d9fe](https://github.com/vitessio/vitess/pull/11909/commits/6b2d9fef24a29caec72c2269e335b209013dfa97)

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+						}
+					}
+
+					printEvents(evs) // for debugging ci failures
```
Moved by [f7e480f](https://github.com/vitessio/vitess/pull/11909/commits/f7e480f7238d6fb2ba9c454609e9d69ef7fc274d)

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+								return completedEvs[i].GetKeyspace() < completedEvs[j].GetKeyspace()
+							}
+							return completedEvs[i].GetShard() < completedEvs[j].GetShard()
+						})
```
It makes sense. This is not identical to the other one, but it can be used for both cases, so I created a utility function by [a013980](https://github.com/vitessio/vitess/pull/11909/commits/a013980946e119147d1a4b68af1f6ca045b1aaa9).

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+							return completedEvs[i].GetShard() < completedEvs[j].GetShard()
+						})
+						for i, ev := range completedEvs {
+							require.Regexp(t, c.expectedCompletedEvents[i], ev.String())
```
I just left an unfinished line of code, so I removed it with the commit [cd119d4](https://github.com/vitessio/vitess/pull/11909/commits/cd119d4bb5cca0ad375f88fb4fb58c784e13549d) 🙇‍♂️

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+						t.Logf("TestVStreamCopyWithoutKeyspaceShard was successful")
+						return
+					} else if c.expectedEventNum < len(evs) {
+						t.Fatalf("len(events)=%v are not expected\n", len(evs))
```
Replaced by [ba8](https://github.com/vitessio/vitess/pull/11909/commits/ba8c231fce80b23bd506a68b97d042075ce6db42)

yoheimuta commented on go/vt/vtgate/endtoend/vstream_test.go:
```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+						t.Fatalf("len(events)=%v are not expected\n", len(evs))
+					}
+				case io.EOF:
+					log.Infof("stream ended\n")
+					cancel()
```
I initially thought that this code was unnecessary and didn't have any meaningful purpose, but I realized that it might have been added for a specific reason based on the assumption that the existing test code was reasonable and robust. To clarify the intent behind this code, I will ask the original implementer in our chat.

- https://github.com/vitessio/vitess/commit/a24770f2089dae3b532ad1ffb77e2a5a47604c7c#diff-c98b8f318aac471272bf650c7824713c9664b46895380acf322ec3498a2fe578R221

yoheimuta replied:
Hi 👋 @rohit-nayak-ps 
I'm wondering if I can get rid of the code that cancels the context if the stream ends unexpectedly. It looks like it's not needed, but this line doesn't run in a normal scenario, so I want to double-check it just in case.

mattlord replied:
I was curious. This isn't blocking.

yoheimuta commented:
@mattlord Thanks for reviewing my work! 
I just finished responding to your comments, so please take a look when you have a chance. 

The only thing left on my to-do list for now is addressing the comment on [#11909](https://github.com/vitessio/vitess/pull/11909#discussion_r1106812847), which I'll bring up in our chat shortly.

yoheimuta commented:
> which I'll bring up in our chat shortly.

Instead, I first posted a question as a follow-up comment.
https://github.com/vitessio/vitess/pull/11909#discussion_r1106871975

yoheimuta commented:
Regarding commit [8970182](https://github.com/vitessio/vitess/pull/11909/commits/89701825241f1fc6eba8ec51eb6b06b789a274a0), a simple change to a test log message and comment unexpectedly caused 17 checks to fail. I believe these failures are unrelated to the changes in the PR, as the previous commit [d7169ec](https://github.com/vitessio/vitess/pull/11909/commits/d7169ec59f3e28e6d0320525a4cf2be3f8878a0c) only failed 1 check.

rohit-nayak-ps commented:
> Regarding commit [8970182](https://github.com/vitessio/vitess/pull/11909/commits/89701825241f1fc6eba8ec51eb6b06b789a274a0), a simple change to a test log message and comment unexpectedly caused 17 checks to fail. I believe these failures are unrelated to the changes in the PR, as the previous commit [d7169ec](https://github.com/vitessio/vitess/pull/11909/commits/d7169ec59f3e28e6d0320525a4cf2be3f8878a0c) only failed 1 check.

The errors are due to a change in the signature of a test function, see: https://github.com/vitessio/vitess/pull/11909/commits/89701825241f1fc6eba8ec51eb6b06b789a274a0

You probably need to `rebase` to `main` if the tests are running locally for you.

yoheimuta commented:
@rohit-nayak-ps Thank you for your information!
I mixed up [GitHub Action](https://github.com/actions/checkout/issues/15#issuecomment-524093065) with [CircleCI](https://discuss.circleci.com/t/checkout-merged-code-for-pull-requests/15398) 😅. 

[256400f](https://github.com/vitessio/vitess/pull/11909/commits/256400fe92a6338085b7cf0ee918ca34c9a0a362) fixed the test signature.

mattlord reviewed (approved):
Thank you, @yoheimuta ! ❤️ This is a nice usability improvement.

mattlord commented:
@yoheimuta if we merge in / rebase on main now, the tests should all be good.

yoheimuta commented:
@mattlord Thank you for your review and helpful suggestions! I really appreciate it.

yoheimuta commented:
Thanks for the info! I have merged the main branch.

yoheimuta commented:
@mattlord Is there anything missing to pass the test?
I don't think the flakiness of [Cluster (onlineddl_vrepl) mysql57](https://github.com/vitessio/vitess/actions/runs/4258358808/jobs/7410030110) and [Cluster (onlineddl_vrepl)](https://github.com/vitessio/vitess/actions/runs/4192255158/jobs/7268653218) are caused by this PR, though.

yoheimuta commented:
And I'm thinking of preparing a PR for https://github.com/vitessio/website after it's been approved by two people.

mattlord commented:
> @mattlord Is there anything missing to pass the test?

I don't think so. We just need @rohit-nayak-ps to take another look when he can (PRs need 2 approvals). 

> I don't think the flakiness of [Cluster (onlineddl_vrepl) mysql57](https://github.com/vitessio/vitess/actions/runs/4258358808/jobs/7410030110) and [Cluster (onlineddl_vrepl)](https://github.com/vitessio/vitess/actions/runs/4192255158/jobs/7268653218) are caused by this PR, though.

Those are unrelated. It's a known issue that we're discussing the proper/best solution to. 

Thanks!

yoheimuta commented:
All tests are passed! Thank you.

rohit-nayak-ps reviewed (approved):
lgtm. Nice work!
